packages/hooks/use-attrs/index.ts

```diff
@@ -0,0 +1,95 @@
+import {
+  getCurrentInstance,
+  onMounted,
+  reactive,
+  shallowRef,
+  watch,
+  watchEffect,
+} from 'vue'
+import { entries, isUndefined, isEqual } from '@element-plus/utils/util'
+import { addClass, removeClass, setStyle, removeStyle } from '@element-plus/utils/dom'
+import type { ComponentInternalInstance } from 'vue'
+
+interface Attrs extends Record<string, unknown> {
+  class?: string
+  style?: Hash<string>
+}
+
+const DEFAULT_EXCLUDE_KEYS = ['class', 'style']
+const LISTENER_PREFIX = /^on[A-Z]/
+
+function syncAttribute<T extends string, P = Attrs[T]>(
+  instance: ComponentInternalInstance,
+  attributeName: T,
+  handler: (instance: ComponentInternalInstance, newVal: P, oldVal?: P) => void,
+) {
+  const attrs = instance.attrs
+  const initialValue = attrs[attributeName]
+
+  if (isUndefined(attrs[attributeName])) return
+
+  onMounted(handler.bind(null, instance, initialValue))
+  watch(
+    () => attrs[attributeName],
+    handler.bind(null, instance),
+  )
+}
+
+function syncClass(
+  instance: ComponentInternalInstance,
+  newClass: string,
+  oldClass?: string,
+) {
+  const el = instance.vnode.el as HTMLElement
+
+  if (!el) return
+
+  oldClass && removeClass(el, oldClass)
+  newClass && addClass(el, newClass)
+}
+
+function syncStyle(
+  instance: ComponentInternalInstance,
+  newStyle: CSSStyleDeclaration,
+  oldStyle?: CSSStyleDeclaration,
+) {
+  const el = instance.vnode.el as HTMLElement
+
+  if (!el || isEqual(newStyle, oldStyle)) return
+
+  oldStyle && removeStyle(el, oldStyle)
+  newStyle && setStyle(el, newStyle)
+}
+
+export default (excludeListeners = false, excludeKeys: string[] = []) => {
+  const instance = getCurrentInstance()
+  const { inheritAttrs } = instance.type
+  const attrs = shallowRef({})
+  const allExcludeKeys = excludeKeys.concat(DEFAULT_EXCLUDE_KEYS)
+
+  // Since attrs are not reactive, make it reactive instead of doing in `onUpdated` hook for better performance
+  instance.attrs = reactive(instance.attrs)
+
+  watchEffect(() => {
+    const res = entries(instance.attrs)
+      .reduce((acm, [key, val]) => {
+        if (
+          !allExcludeKeys.includes(key) &&
+          !(excludeListeners && LISTENER_PREFIX.test(key))
+        ) {
+          acm[key] = val
+        }
+
+        return acm
+      }, {})
+
+    attrs.value = res
+  })
```

Comment:
To me this seems a little overkill for this feature, because you have you idea when the `attrs` gets changed, which means you run this whenever the component gets updated.

Reply:
IMHO holding a cache of `attrs` to avoid unnecessary updates could solve the problem. 
BTW, since you have made `attrs` reactive, can you use `watch(() => attrs)` that only updates when `attrs` gets changed?
Also, I think at this particular point, you could set `instance.type.inheritAttrs` to `false` to avoid automatically inheritance to `root element` 
These were all based on my understanding, correct me when you feel like you should.

Reply:
You may have some misunderstanding of `watchEffect` , it only calls when dependencies change but not whenever component updates, so `watchEffect` could make same effect as `watch(() => attrs)`. And I'm afraid that I can't argree with you about modifying `instance.type.inheritAttrs`, cause it's kind of impure and unnecessary.